### ISSUE TITLE
srand() behavior with large numbers

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8650]
Migrated from [rt.perl.org#40605](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=40605) (status was 'resolved')

Searchable as RT40605$

**Comment by p5pRT** [comment]
### From ski-perlbug@allafrica.com

#### Created by ski-perlbug@allafrica.com


This is a bug report for perl from ski\-perlbug@&#8203;allafrica\.com\,
generated with the help of perlbug 1\.35 running under perl v5\.8\.7\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

When srand is given a number that is not representable as a native 
integer\, it produces the same results\. While understandable\, it is
trivial to fix \(see below\)&#8203;:

Here is the current output\, which I would argue is bogus&#8203;:
ski@&#8203;chronicle&#8203;:\~$ perl \-e 'foreach my $i \(\-2\.\.2\) \{ srand\(\-2\*\*31\+$i\); print "$i&#8203;: "\, rand\, "\\n"; \}'
\-2&#8203;: 0\.67082803610629
\-1&#8203;: 0\.67082803610629
0&#8203;: 0\.67082803610629
1&#8203;: 0\.541630344771878
2&#8203;: 0\.412432653437467
ski@&#8203;chronicle&#8203;:\~$ perl \-e 'foreach my $i \(\-2\.\.2\) \{ srand\(2\*\*32\+$i\); print "$i&#8203;: "\, rand\, "\\n"; \}'
\-2&#8203;: 0\.429223418775113
\-1&#8203;: 0\.300025727440701
0&#8203;: 0\.300025727440701
1&#8203;: 0\.300025727440701
2&#8203;: 0\.300025727440701

I should never be able to get the same sequence from&#8203;:
&nbsp;  srand\(foo\(\)\+1\)
and
&nbsp;  srand\(foo\(\)

It seems that perl should handle this situation by using the moral 
equivalent of this code&#8203;:
sub srand\_wrapper \{
&nbsp;    \# $\_\[0\] % \*2\*\*\(size\_in\_bits\(int\)\-1\)\)\,
&nbsp;   \# e\.g\. $\_\[0\] % 2\*\*31 on 32 bit machines
&nbsp;    \#      $\_\[0\] % 2\*\*63 on 64 bit machines\, etc\.
&nbsp;    return srand\($\_\[0\] % \(1\<\<\(\(length\(pack\("i"\, 1\)\)\<\<3\)\-1\)\)\)
\};

The desired behavior\, such as that derived by using the above function\, 
is&#8203;:
ski@&#8203;chronicle&#8203;:\~$ perl \-e 'sub srand\_wrapper \{ return srand\($\_\[0\] % \(1\<\<\(\(length\(pack\("i"\, 1\)\)\<\<3\)\-1\)\)\) \}; foreach my $i \(\-2\.\.2\) \{ srand\_wrapper\(\-2\*\*31\+$i\); print "$i&#8203;: "\, rand\, "\\n" \}'
\-2&#8203;: 0\.929223418775113
\-1&#8203;: 0\.800025727440701
0&#8203;: 0\.17082803610629
1&#8203;: 0\.0416303447718782
2&#8203;: 0\.912432653437467
ski@&#8203;chronicle&#8203;:\~$ perl \-e 'sub srand\_wrapper \{ return srand\($\_\[0\] % \(1\<\<\(\(length\(pack\("i"\, 1\)\)\<\<3\)\-1\)\)\) \}; foreach my $i \(\-2\.\.2\) \{ srand\_wrapper\(2\*\*31\+$i\); print "$i&#8203;: "\, rand\, "\\n" \}'
\-2&#8203;: 0\.929223418775113
\-1&#8203;: 0\.800025727440701
0&#8203;: 0\.17082803610629
1&#8203;: 0\.0416303447718782
2&#8203;: 0\.912432653437467

Note the property that 
&nbsp;  srand\(2\*\*\(int\_size\_in\_bits\-1\)\+$n\)
is the same as calling&#8203;:
&nbsp;  srand\(\-2\*\*\(int\_size\_in\_bits\-1\)\+$n\)
\.\.\. this was already the case\, e\.g\.
&nbsp;  srand\(2\*\*\(int\_size\_in\_bits\-1\)\+$i\) == srand\(\-2\*\*\(int\_size\_in\_bits\-1\)\+$i\)
\.\.\.but is now more visible\, because values > 2\*\*\(int\_size\_in\_bits\-1\) 
don't wrap and become negative anymore\. So\, the random streams we get 
when we use numbers between 2\*\*\(int\_size\_in\_bits\-1\) and 
2\*\*int\_size\_in\_bits will be different\. If this is something we want to 
worry about\, the current behavior could be preserved with a bit more 
effort which I don't have time to flesh out right now \.\.\. but obviously 
this all should be done in C\, not perl\, and I have not the chops to do 
so\. Tested on linux/5\.8\.7\, linux/5\.8\.4\, and macosx/5\.8\.8 \.

Cheers\,
Brian Szymanski
ski\-perlbug@&#8203;allafrica\.com

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.7:

Configured by Debian Project at Fri Dec 16 08:00:01 UTC 2005.

Summary of my perl5 (revision 5 version 8 subversion 7) configuration:
  Platform:
    osname=linux, osvers=2.6.10, archname=i486-linux-gnu-thread-multi
    uname='linux rothera 2.6.10 #1 smp fri may 13 09:24:22 utc 2005 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i486-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8 -Darchlib=/usr/lib/perl/5.8 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.7 -Dsitearch=/usr/local/lib/perl/5.8.7 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.7 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.0.3 20051204 (prerelease) (Ubuntu 4.0.2-5ubuntu2)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.3.5.so, so=so, useshrplib=true, libperl=libperl.so.5.8.7
    gnulibc_version='2.3.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    SPRINTF0 - fixes for sprintf formatting issues - CVE-2005-3962


@INC for perl v5.8.7:
    /etc/perl
    /usr/local/lib/perl/5.8.7
    /usr/local/share/perl/5.8.7
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8
    /usr/share/perl/5.8
    /usr/local/lib/site_perl
    /usr/local/lib/perl/5.8.4
    /usr/local/share/perl/5.8.4
    .


Environment for perl v5.8.7:
    HOME=/home/ski
    LANG=en_US.UTF-8
    LANGUAGE=en
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/bin/X11:/usr/games:/home/ski/bin:/usr/local/sbin:/usr/sbin:/sbin:/var/qmail/bin:/allafrica/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From Roch.Skelton@marriott.com

#### Created by Roch.Skelton@marriott.com



srand appears to fail if the EXPR given is "too big"

I'm trying to use a better "unique" value for the seed than the current
time\,
specifically the count of bytes on the ethernet adapter\. Using just the
inbound
or just the outbound byte count gives me valid "random" results\.
However\, when
I concatenate these 2 values as my EXPR for srand\, rand\(\) returns the
exact same
result every time\. My assumption is that srand overflows and actually
seeds with
a "1" or zero or some other constant value\. I do not need to merge both
values
together\, but I do need to know the maximum value that can be used by
srand\, so
I can truncate the byte count if even one value is "too big"\. \(I know
that a 10
digit EXPR works fine\, and a 20 digit EXPR is too big\, just not sure
where the
line is in between these\.\)

Thanks for any clarification to this "feature"\.

Roch Skelton
Marriott International

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.2:

Configured by ActiveState at Tue Dec  9 10:19:27 2003.

Summary of my perl5 (revision 5 version 8 subversion 2) configuration:
  Platform:
    osname=MSWin32, osvers=4.0, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=undef use5005threads=undef useithreads=define
usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -Gf -W3 -MD -Zi -DNDEBUG -O1 -DWIN32
-D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT  -DNO_HASH_SEED
-DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO
-DPERL_MSVCRT_READFIX',
    optimize='-MD -Zi -DNDEBUG -O1',
    cppflags='-DWIN32'
    ccversion='', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf
-libpath:"C:\Perl\lib\CORE"  -machine:x86'
    libpth=C:\PROGRA~1\MICROS~3\VC98\lib
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib
netapi32.lib uuid.lib wsock32.lib mpr.lib winmm.lib  version.lib
odbc32.lib odbccp32.lib msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib
winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib
oleaut32.lib  netapi32.lib uuid.lib wsock32.lib mpr.lib winmm.lib
version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl58.lib
    gnulibc_version='undef'
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug
-opt:ref,icf  -libpath:"C:\Perl\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY
    21846 Configure gets d_u32align wrong
    21739 [perl #24493] install.html not working
    21737 Ooops. left an XXX comment in, and worse still it's a //
comment
    21735 utf8 keys now work for tied hashes
    21734 Accessing unicode keys in tie hashes via hv_exists was broken
    21733 ext/threads/t/problem.t
    21732 Config::myconfig() fails under ithreads
    21728 Update perlhist with 5.6.2
    21723 Include 'SCCS' in the list of dir names ignored by installperl
    21718 Empty subroutine as object method segfaults in 5.8.2
(sometimes)
    21714 Fix bug #24380: assigning list with duplicated keys to a hash
    21706 [perl #24460] [DOC PATCH] the begincheck program
    21693 must copy changes from win32/makeifle.mk to wince/makefile.ce
    21691 Update the list of pumpkings in perlhist.pod
    21687 [PATCH 5.6.2-RC1 pod/perlhist.pod]  Updated
    21677 OS/2 docu
    21676 Bug #24407: key for shared hash got stringified into wrong
pool
    21673 Be sure to use -fPIC not -fpic on Linux/SPARC
    21672 extending the hash attack test
    21671 Benchmark.pm cmpthese segfault
    21662 'make minitest' fails for op/cproto and op/pat
    21586 Comment that this 'optimisation' is actually a necessary fixup
    21548 Sync with Pod::Perldoc 3.12
    21540 Fix backward-compatibility issues in if.pm


@INC for perl v5.8.2:
    C:/Perl/lib
    C:/Perl/site/lib
    .


Environment for perl v5.8.2:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
 
PATH=C:\Perl\bin\;C:\MICROS\Res\Pos\Bin\;C:\MICROS\Common\Bin\;C:\WINDOW
S\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\Program
Files\Sybase\ASA\shared\win32\;C:\Program
Files\Sybase\ASA\Win32;C:\Program Files\Common Files\Crystal
Decisions\2.0\Bin\;C:\Program Files\Symantec\pcAnywhere\
    PERL_BADLANG (unset)
    SHELL (unset)

--- also fails on ---

Site configuration information for perl v5.6.1:

Configured by ActiveState at Tue Apr 13 19:24:10 2004.

Summary of my perl5 (revision 5 version 6 subversion 1) configuration:
  Platform:
    osname=MSWin32, osvers=4.0, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=undef use5005threads=undef useithreads=define
usemultiplicity=define
    useperlio=undef d_sfio=undef uselargefiles=undef usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cl', ccflags ='-nologo -O1 -MD -Zi -DNDEBUG -DWIN32 -D_CONSOLE
-DNO_STRICT -DHAVE_DES_FCRYPT  -DPERL_IMPLICIT_CONTEXT
-DPERL_IMPLICIT_SYS -DPERL_MSVCRT_READFIX',
    optimize='-O1 -MD -Zi -DNDEBUG',
    cppflags='-DWIN32'
    ccversion='', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=4
    alignbytes=8, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf
-libpath:"D:\Perl\lib\CORE"  -machine:x86'
    libpth="D:\Perl\lib\CORE"
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib
netapi32.lib uuid.lib wsock32.lib mpr.lib winmm.lib  version.lib
odbc32.lib odbccp32.lib msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib
winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib
oleaut32.lib  netapi32.lib uuid.lib wsock32.lib mpr.lib winmm.lib
version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl56.lib
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug
-opt:ref,icf  -libpath:"D:\Perl\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY


@INC for perl v5.6.1:
    D:/Perl/lib
    D:/Perl/site/lib
    .


Environment for perl v5.6.1:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/usr/lib:/usr/X11R6/lib
    LOGDIR (unset)
    PATH=D:\Perl\bin\;C:\Program Files\Support Tools\;C:\Program
Files\Windows Resource
Kits\Tools\;D:\micros\MSSQL\Backup\BackupUtilities;;D:\MICROS\LES\POS\97
00\configurator;;D:\MICROS\LES\POS\9700\scripts;;D:\MICROS\LES\POS\9700\
etc;;D:\MICROS\LES\POS\9700\bin;;C:\cygwin\bin;C:\WINDOWS\system32;C:\WI
NDOWS;C:\WINDOWS\System32\Wbem;C:\Program Files\Microsoft SQL
Server\80\Tools\BINN;C:\SFU\common\;C:\Program Files\Microsoft SQL
Server\90\Tools\binn\
    PERL_BADLANG (unset)
    SHELL (unset)



```
</details>

**Comment by p5pRT** [comment]
### From @rgs

2008/12/8 via RT Skelton\, Roch \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> srand appears to fail if the EXPR given is "too big"
>
> I'm trying to use a better "unique" value for the seed than the current
> time\,
> specifically the count of bytes on the ethernet adapter\. Using just the
> inbound
> or just the outbound byte count gives me valid "random" results\.
> However\, when
> I concatenate these 2 values as my EXPR for srand\, rand\(\) returns the
> exact same
> result every time\. My assumption is that srand overflows and actually
> seeds with
> a "1" or zero or some other constant value\. I do not need to merge both
> values
> together\, but I do need to know the maximum value that can be used by
> srand\, so
> I can truncate the byte count if even one value is "too big"\. \(I know
> that a 10
> digit EXPR works fine\, and a 20 digit EXPR is too big\, just not sure
> where the
> line is in between these\.\)

Perl calls the random number generator provided by your system's C library\.
The C type corresponding to the seed can be found out by

&nbsp;    perl \-V&#8203;:randseedtype

So\, depending on your platform\, you can know where it overflows\.
\(On my Linux it's "long"\)\.

> Thanks for any clarification to this "feature"\.

Perl's srand\(\) should probably warn in case of overflow\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From module@renee-baecker.de

There is a similar 2\-year\-old ticket&#8203;:
http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=40605

**Comment by p5pRT** [comment]
### From @doy

Fixed in f914a68\. There's not really a good way to do anything sensible
for this case\, since perl doesn't have built\-in arbitrary precision
arithmetic\, but I've added a warning for the case where it does overflow\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Fixed in f914a68\. There's not really a good way to do anything sensible
for this case\, since perl doesn't have built\-in arbitrary precision
arithmetic\, but I've added a warning for the case where it does overflow\.

**Comment by p5pRT** [comment]
@doy - Status changed from 'new' to 'resolved'
